### PR TITLE
refactor(ci): use docker instead of Konlet for GCP deployments in CI

### LINKED ISSUE #4080: ci: stop re-checking the same Google Cloud log line in a loop
## Motivation

In some workflows, we re-check the same Google Cloud log line multiple times in a loop:
https://github.com/ZcashFoundation/zebra/blob/6d2c4fbb5f112abea6528984eab14349f3548584/.github/workflows/test-full-sync.yml#L238-L242

This log line will never change, because the `zebrad` process has finished.

This could cause hangs, but it doesn't seem to be doing that right now, so it's a low priority.

### Designs

In every workflow that checks logs after `zebrad` has exited:
- Just check the log lines once
- If that causes failures, go back further in the logs

### LINKED ISSUE #4090: ci: log errors caused by deployment or infrastructure issues
## Motivation

Sometimes CI fails because of a deployment or infrastructure issue, but this issues are hard to track if we don't show those in the logs, at least to know that the same issue is causing multiple checks to fail, like the one from the screenshot below, which has been causing multiple failures in the past week, and re-running after an specific time fixes the issue

image

Examples:
- https://github.com/ZcashFoundation/zebra/runs/5964831122?check_suite_focus=true#step:9:115
- https://github.com/ZcashFoundation/zebra/runs/5965260522?check_suite_focus=true#step:9:115

### Designs
- If reading the container logs fails, add further information to the PR (annotations?) to understand what's causing it

### LINKED ISSUE #4114: ci: GCP workflows fail with missing container names, due to mounting cached disk partition
## Motivation

Some Zebra workflows repeatedly fail due to a missing container error:
> Error: No such container: klt-regenerate-disk-4038-merge-9969dc2-bfnh

https://github.com/ZcashFoundation/zebra/runs/6007481272?check_suite_focus=true#step:10:115

But the container actually exists - it's the partitions on the image that trigger a bug in Google Cloud.

### Underlying Infrastructure Bug

TODO: @gustavovalverde knows the details here.

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

This is a workaround for an issue related to disk partitioning, caused by a GCP service called Konlet, while mounting the cached disks to the VM and then to the container.

This has been causing most integration tests to fail when a disk with cached state is needed.

Fixes #4114
Closes #4080
Closes #4090
Depends-On: #4266

## Solution

- Do not run our tests directly by GCP services, instead  manually create a container inside the VM with docker

** Drawbacks **
- We won't have the container logs available in GCP, but this is not as important as having the logs available in GitHub Actions, and we're not yet using GCP for reporting.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

Follow up with GCP support ticket to confirm if there's a fix or an unknown workaround which doesn't require us to stop using Konlet.

### CODE REVIEW

**Comment by gustavovalverde** [comment]
I've confirm we'll have to use this workaround. The solutions given by GCP's support (which fixed other issues) didn't fix this one.

File: .github/workflows/test-full-sync.yml [hunk 1]
```diff
@@ -181,21 +158,24 @@ jobs:
...
+          gcloud compute ssh \
+          full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
+          --zone ${{ env.ZONE }} \
+          --quiet \
+          --ssh-flag="-o ServerAliveInterval=5" \
+          --command="docker logs full-sync --tail 20")
+
+          SYNC_HEIGHT=$(echo $DOCKER_LOGS | grep -oE 'finished initial sync to chain tip, using gossiped blocks sync_percent=100.000 % current_height=Height\([0-9]+\)' | grep -oE '[0-9]+' | tail -1 || [[ $? == 1 ]])
 
-          echo "SYNC_HEIGHT=$SYNC_HEIGHT" >> $GITHUB_ENV
```
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Possible Bug: Does the `SYNC_HEIGHT` environmental variable carry through to the next step without `>> $GITHUB_ENV`?

File: docker/Dockerfile [hunk 2]
```diff
@@ -78,7 +78,7 @@ FROM builder AS tester
 COPY --from=us-docker.pkg.dev/zealous-zebra/zebra/zcash-params /root/.zcash-params /root/.zcash-params
 COPY --from=us-docker.pkg.dev/zealous-zebra/zebra/lightwalletd /lightwalletd /usr/local/bin
 
-RUN cargo test --locked --release --features enable-sentry --workspace --no-run
+RUN cargo test --locked --release --features enable-sentry --no-run
```
Comment (teor2345) [inline docker/Dockerfile]: Question: why is this change in this PR? What does it do?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -203,76 +203,42 @@ jobs:
         id: create-instance
         if: ${{ steps.changed-files-specific.outputs.any_changed == 'true' || github.event.inputs.regenerate-disks == 'true' || github.event_name == 'push'}}
         run: |
-          gcloud compute instances create-with-container "regenerate-disk-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}" \
+          gcloud compute instances create-with-container "sync-to-checkpoint-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }}" \
           --boot-disk-size 100GB \
           --boot-disk-type pd-ssd \
-          --create-disk name="zebrad-cache-${{ env.GITHUB_SHA_SHORT }}-${{ env.NETWORK }}-checkpoint",size=100GB,type=pd-ssd \
```
Comment (teor2345) [inline .github/workflows/test.yml]: Functional Change: This change drops the Zcash network from the disk name.

That's ok for now, but I'd like to document that the `NETWORK` env var can't be changed:
- the tests and images only work on mainnet
- a TODO for ticket #1791 (or a new ticket for running tests on testnet)

Or we could just remove the network env var entirely.

File: zebra-state/src/constants.rs [hunk 4]
```diff
@@ -31,3 +31,4 @@ lazy_static! {
     /// Regex that matches the RocksDB error when its lock file is already open.
     pub static ref LOCK_FILE_ERROR: Regex = Regex::new("(lock file).*(temporarily unavailable)|(in use)|(being used by another process)").expect("regex is valid");
 }
+
```
Comment (teor2345) [inline zebra-state/src/constants.rs]: Workflow triggers: Does PR #4237 avoid you having to make unrelated changes to Rust files to test this PR? Or do we need to make more workflow trigger fixes?

**Review by teor2345 — COMMENTED** [review]
This is a large PR, which I'm having trouble understanding and reviewing.

It looks like the diff contains:
- changes from previous PRs in this series
- bug fixes
- functional changes
- syntax and name cleanups

I'd really like to see:
- a list of changes in this PR
- how you've tested it
- test logs that show that we haven't accidentally skipped any tests

If it's not possible to test it until it is merged to `main`, can we just fix one workflow or disk first? Then we can fix the other workflows once we know the fix works.

File: .github/workflows/test.yml (hunk 3 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml]: I know this is confusing, but this change only drops the `NETWORK` from the disk used by the instance, not by the image (which is the one holding the cache). The image is built here https://github.com/ZcashFoundation/zebra/pull/4252/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R278

- I've assigned myself to: https://github.com/ZcashFoundation/zebra/issues/1791 and also added to the next sprint, to fix all issues related to Testnet (this is not one, but there are other places with a hardcoded Mainnet)

Let's track that ticket to solve this in another PR.

File: zebra-state/src/constants.rs (hunk 4 above)
Comment (gustavovalverde) [reply zebra-state/src/constants.rs]: We need to make more workflow trigger fixes:

I'm removing this change in the meanwhile. I thought of a solution for this triggers. I'll test that in another PR.

Fixed by: [fix: revert uneeded changes](https://github.com/ZcashFoundation/zebra/pull/4252/commits/f516ae5b636a84b20f70444847fe86ef8b6d551b)

File: docker/Dockerfile (hunk 2 above)
Comment (gustavovalverde) [reply docker/Dockerfile]: Fixed by: [fix: revert uneeded changes](https://github.com/ZcashFoundation/zebra/pull/4252/commits/f516ae5b636a84b20f70444847fe86ef8b6d551b)

File: .github/workflows/test-full-sync.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/test-full-sync.yml]: Oops: [fix(ci): revert wrong deletion](https://github.com/ZcashFoundation/zebra/pull/4252/commits/dc0bb66f6e772ee263b0f13d7c766b4d976e43c8)

File: .github/workflows/test.yml (hunk 3 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml]: As this is not required (at least in this PR) I'm reducing the diff

**Comment by gustavovalverde** [comment]
@teor2345  I've removed everything that's not related to this PR

**Comment by gustavovalverde** [comment]
After merging
-  https://github.com/ZcashFoundation/zebra/pull/4266

This is the most relevant PR to get `lightwalletd` CI working

File: .github/workflows/test.yml (hunk 3 above)
Comment (teor2345) [reply .github/workflows/test.yml]: > I've assigned myself to: #1791 and also added to the next sprint

We can't do reliable testnet tests, because we decided not to fix bugs that are blocking them. So that ticket should have been closed. (I closed it.)

Right now our top priority is getting CI working, and getting the lightwalletd tests working in CI.

Would it help if I stopped mentioning unrelated bugs?

**Review by teor2345 — APPROVED** [review]
Alright, let's try this and see how it goes.

**Review by dconnolly — COMMENTED** [review]
👍
